### PR TITLE
opt: FK checks for Upsert

### PR DESCRIPTION
**Pull request by RaduBerinde**
These commit add support for opt-driven FK checks for UPSERT and INSERT ON CONFLICT DO UPDATE. This is a simplified and cleaned up reimplementation of #45095.

There aren't a lot of logictests around FKs and upsert. I plan to try to build a randomized testing facility for FK checks, where we generate randomized mutations and cross-check FK violations or lack there of against a separate instance of the database without FK relationships.

#### Revert "opt: add first half of upsert FK checks"

This reverts commit 96447c8d2e11ad888d8c39b990c33fea18f4ed9f.

We are reconsidering the direction of the upsert FK checks: instead of filtering
inserted vs updated rows, we can use the return columns to get the "new" values.
This change also made some things harder to understand: we used to build the FK
check input using just the columns corresponding to the FK instead of all
columns.

Release note: None

#### opt: refactor optbuilder FK check code

The optbuilder FK check code has become unnecessarily complicated:
 - there are two ways of creating a WithScan (`makeFKInputScan` and
   `projectOrdinals`) which are similar but take different kinds of information
   as input.
 - there are various slices of column ordinals or column IDs that are threaded
   through long parts of code, making it hard to follow.

This change cleans this up by creating a fkCheckHelper which contains the
metadata related to FK table ordinals and is capable of creating a WithScan with
either the "new" or the "fetched" values.

The new code should generate the same effective plans; the differences are:
 - we now always create the WithScan before the other table Scan so some column IDs in the plan changed;
 - we no longer have an unnecessary projection for Update (it was used to
   renumber the columns coming out of WithScan).

Release note: None

#### opt: add outbound FK checks for upsert

This is a re-implementation of Justin's change 3d1dd0fdfb, except that we now
perform the insertion check for all new rows instead of just the inserted rows.
We do this by utilizing the same column that would be used for a RETURNING
clause, which in some cases is of the form
`CASE WHEN canary IS NULL col1 ELSE col2`.

Release note: None

#### opt: add inbound FK checks for upsert

This change adds inbound FK checks for upsert and switches execution over to the
new style FK checks for upsert.

Similar to UPDATE, the inbound FK checks run on the set difference between "old"
values for the FK columns and "new" values.

Release note (performance improvement): improved execution plans of foreign key
checks for UPSERT and INSERT ON CONFLICT in some cases (in particular
multi-region).

### CODE REVIEW

**Review by andy-kimball — COMMENTED** [review]
Is there a comment block somewhere that gives an overview of FK checks for each of the operators (Insert/Update/Delete/Upsert)? I'm looking for something to give me overall context on how we compile these things. An example is the write-up I did for `Builder.buildInsert` and `Builder.buildUpdate`, etc.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1C2WDDC2cuHgKhzc0x:bb38knn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @rytaft)

**Review by RaduBerinde — COMMENTED** [review]
Added something, see if this helps. I can add more details / examples if you think it helps.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1C8tyN4E-_Lz-YdoEq:b-yyej0v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

**Review by andy-kimball — COMMENTED** [review]
A top-level question: have you thought through the impact of mutation columns (i.e. cols in process of being added or removed from table) on FK checks? Are there any tests around this that would be good? See the `optbuilder/testdata/insert` `mutation` table tests for examples. Might also be interesting to have a few logic test cases around this. See `subtest regression_35611` for good way to test that kind of case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1CzUUC8DuQ6m73LIJu:bskzpru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/fk_opt, line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1DLvJA91s1c5zhA00y:-M1DLvJB-D2i2vLJs-GN:b-hnk0nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/logictest/testdata/logic_test/fk_opt#L250)):*
> ```
> 
> statement ok
> DROP TABLE parent
> ```

Do we have good self-ref cases already for UPSERT + FK? If not, I'd add some.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 886 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1CvVgzBkIQ2Jw9hmi0:-M1CvVgzBkIQ2Jw9hmi1:bkyi2hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L886)):*
> ```Go
> // In the case of insert, each FK check query is an anti-join with the left side
> // being a WithScan of the mutation input and the right side being the
> // referenced table.
> ```

I think a small example here would be useful, or maybe a reference to a good test case (like `fk-checks-insert`).
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1031 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1D09aZ68-kQdMVe9JG:-M1D09aZ68-kQdMVe9JH:bk6qh8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1031)):*
> ```Go
> 
> 		// Construct an Except expression for the set difference between "old"
> 		// FK values and "new" FK values.
> ```

Can you add more explanation/examples that show the need for `Except` here? I had to spend time pondering the "whys", including why we need a set operator here but not in the insert case above. What's so special about this check, I asked myself, since my intuition incorrectly suggested that the incoming and outgoing cases would be symmetric. Also, I wondered why’s it'd be bad to just make deletion checks for all the rows. Was this just a performance check, or was it a semantic necessity? These are the kinds of questions you want to answer in commentary.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1147 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1D7Hhe1UDcaY9DoLJz:-M1D7Hhe1UDcaY9DoLK-:blan4sm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1147)):*
> ```Go
> // The input to the insertion check will be produced from the input to the
> // mutation operator.
> func (mb *mutationBuilder) addInsertionCheck(fkOrdinal int) {
> ```

NIT/Question: How come `addInsertionCheck` and `addDeletionCheck` are in `mutationBuilder` rather than in `fkCheckHelper`? I was a bit surprised to see the input scans being created in the helper but the join being created here.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1376 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1DIl1HABbvApxnnExJ:-M1DIl1K8hapaV7r7l95:bdzrk55) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1376)):*
> ```Go
> // build the "leaves" of a FK check expression, namely the WithScan of the
> // mutation input and the Scan of the other table.
> type fkCheckHelper struct {
> ```

Should this be in its own file?

**Review by andy-kimball — APPROVED** [review]
:lgtm:, a ton of good work here. Thanks for putting this all together; I'm sure it took a lot of time and detail work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1DMDqkDqLCh9AySzjw:biarfpk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

**Review by RaduBerinde — COMMENTED** [review]
TFTR!

I don't know of a way to have FKs involving mutation columns. You can't add a column and a FK constraint on it at the same time. When deleting a column, the FK constraint gets removed first.

Note that r6 is #45597

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1QfzTT3uZCGYK4tPEd:blhvcdw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/fk_opt, line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1DLvJA91s1c5zhA00y:-M1QkdmW7gBl-Rb5XrEz:beo0owg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/logictest/testdata/logic_test/fk_opt#L250)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Do we have good self-ref cases already for UPSERT + FK? If not, I'd add some.
</blockquote></details>

Good idea. This actually found a problem which I pulled out in #45597.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 886 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1CvVgzBkIQ2Jw9hmi0:-M1SjB6BAE7H0FQfrex2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L886)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think a small example here would be useful, or maybe a reference to a good test case (like `fk-checks-insert`).
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1031 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1D09aZ68-kQdMVe9JG:-M1SohzWAxkdi6PL4Qbc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1031)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you add more explanation/examples that show the need for `Except` here? I had to spend time pondering the "whys", including why we need a set operator here but not in the insert case above. What's so special about this check, I asked myself, since my intuition incorrectly suggested that the incoming and outgoing cases would be symmetric. Also, I wondered why’s it'd be bad to just make deletion checks for all the rows. Was this just a performance check, or was it a semantic necessity? These are the kinds of questions you want to answer in commentary.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1147 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1D7Hhe1UDcaY9DoLJz:-M1Saa-o3BSbtlgDBe3l:b-ooiaof) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1147)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT/Question: How come `addInsertionCheck` and `addDeletionCheck` are in `mutationBuilder` rather than in `fkCheckHelper`? I was a bit surprised to see the input scans being created in the helper but the join being created here.
</blockquote></details>

It's a good point; I didn't want to move the code for these so the diffs are visible. I'd rather make this reorganization in another PR.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1376 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1DIl1HABbvApxnnExJ:-M1S_ziEEqZJCoew1pmv:bw6wmbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d6e3f50960277f81d7ec92f3827a94ff5c8a681/pkg/sql/opt/optbuilder/mutation_builder.go#L1376)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should this be in its own file?
</blockquote></details>

I do want to pull out the FK code to another file but I didn't want to do it as part of this change. I will do it in a follow-up.

**Review by rytaft — APPROVED** [review]
:lgtm:  nice work!

> Reviewed 12 of 12 files at r1, 1 of 7 files at r2, 21 of 21 files at r6, 12 of 12 files at r7, 8 of 8 files at r8, 3 of 3 files at r9, 9 of 9 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-:-M1WKQe_9OZi3QgfZgD3:b-8vtfcy)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 2099 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1W2P1A3e9gsFMcv6Tm:-M1W2P1A3e9gsFMcv6Tn:b-vpajf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7383e9dcd342f304376128718f7258129d5af9d7/pkg/sql/opt/norm/custom_funcs.go#L2099)):*
> ```Go
> 		if !scalarProps.IsAvailable(props.WithUses) {
> 			scalarProps.Shared.Rule.WithUses = c.deriveWithUses(r)
> 			scalarProps.SetAvailable(props.WithUses)
> ```

[nit] this is bothering me that these two lines are in a different order than they are in the previous case. Also why does this case have a comment about lazy calculation and the one above doesn't?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1329 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1W8Nuf3uQWu_gy4Hpo:-M1W8Nuf3uQWu_gy4Hpp:b-rvkjfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49226019989cb79555384a507d6e1b3d9eb215/pkg/sql/opt/optbuilder/mutation_builder.go#L1329)):*
> ```Go
> 		panic(err)
> 	}
> 	// We need SELECT privileges on the referenced table.
> ```

[nit] referenced -> origin
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1160 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WHB8B8D_GwvBb9qFy:-M1WHB8B8D_GwvBb9qFz:be9rn7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1160)):*
> ```Go
> // available, whereas for upsert, the "new" values can be the result of an
> // expression of the form:
> //   CASE WHEN canary IS NULL THEN ELSE updated-value END
> ```

missing something after THEN
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1165 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WHYyB9bomL7UHDG5R:-M1WHYyB9bomL7UHDG5S:bsbibq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1165)):*
> ```Go
> //
> // Only FK relations that involve updated columns result in deletion-side FK
> // check. The insertion-side FK checks are always needed (similar to insert)
> ```

[nit] check -> checks
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1212 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WIF96-jsGer5VYkcS:-M1WIF97-Kzv_-neihee:b-qfbp97) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1212)):*
> ```Go
> 		// Note that technically, to get the old rows we should be selecting only
> 		// the rows that are being updated (using a "canaryCol IS NOT NULL"
> 		// condition); but these rows are harmless because they have all-null
> ```

[nit] "these rows" is a bit vague -- which rows specifically?

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 2099 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1W2P1A3e9gsFMcv6Tm:-M1WZk8KEgVFKm-0jkvK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7383e9dcd342f304376128718f7258129d5af9d7/pkg/sql/opt/norm/custom_funcs.go#L2099)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] this is bothering me that these two lines are in a different order than they are in the previous case. Also why does this case have a comment about lazy calculation and the one above doesn't?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1329 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1W8Nuf3uQWu_gy4Hpo:-M1W_ohS8r8YdKeNGrCG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49226019989cb79555384a507d6e1b3d9eb215/pkg/sql/opt/optbuilder/mutation_builder.go#L1329)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] referenced -> origin
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1160 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WHB8B8D_GwvBb9qFy:-M1WZrJ6CnxUsY9lL5CU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1160)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

missing something after THEN
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1165 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WHYyB9bomL7UHDG5R:-M1WZu0TDoBNuPtt_Ivt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1165)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] check -> checks
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1212 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/45520#-M1WIF96-jsGer5VYkcS:-M1W_Ja-7VujGHv0oqL0:bvqko8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b1c0a365a7e909345d8f46246f2c4764cf7f8/pkg/sql/opt/optbuilder/mutation_builder.go#L1212)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] "these rows" is a bit vague -- which rows specifically?
</blockquote></details>

Good point, see if this is better.

**Review by rytaft — APPROVED** [review]
> Reviewed 2 of 2 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45520)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball)
